File: src/services/articleService.ts
import { get, post, put, del } from './api';
import type { 
  ArticleResponse, 
  ArticleRequest, 
  Page, 
  SearchParams,
  ArticleStatus 
} from '../types/api';

// =============================================================================
// SERVICE DES ARTICLES - ENDPOINTS REST
// =============================================================================

/**
 * Service des articles pour tous les endpoints REST
 * Correspond exactement au contrôleur ArticleController du backend
 */
export const articleService = {
  
  // ---------------------------------------------------------------------------
  // ENDPOINTS PUBLICS (SANS AUTHENTIFICATION)
  // ---------------------------------------------------------------------------
  
  /**
   * Récupère les 10 derniers articles publiés pour la page d'accueil
   * GET /api/articles/recent
   */
  async getRecentArticles(): Promise<ArticleResponse[]> {
    try {
      const articles = await get<ArticleResponse[]>('/api/articles/recent');
      console.log('✅ Recent articles retrieved:', articles.length);
      return articles;
      
    } catch (error) {
      console.error('❌ Failed to get recent articles:', error);
      throw error;
    }
  },

  /**
   * Récupère un article spécifique par son ID (UUID)
   * GET /api/articles/{id}
   */
  async getArticleById(id: string): Promise<ArticleResponse> {
    try {
      const article = await get<ArticleResponse>(`/api/articles/${id}`);
      console.log('✅ Article retrieved:', article.title);
      return article;
      
    } catch (error) {
      console.error(`❌ Failed to get article ${id}:`, error);
      throw error;
    }
  },

  /**
   * Récupère les articles publiés avec pagination
   * GET /api/articles/published?page=0&size=5&sort=publishedAt,desc
   */
  async getPublishedArticles(
    page: number = 0, 
    size: number = 5, 
    sort: string = 'publishedAt,desc'
  ): Promise<Page<ArticleResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        size: size.toString(),
        sort,
      });

      const articlesPage = await get<Page<ArticleResponse>>(
        `/api/articles/published?${params}`
      );
      
      console.log('✅ Published articles retrieved:', {
        totalElements: articlesPage.totalElements,
        currentPage: articlesPage.number,
        totalPages: articlesPage.totalPages,
      });
      
      return articlesPage;
      
    } catch (error) {
      console.error('❌ Failed to get published articles:', error);
      throw error;
    }
  },

  /**
   * Récupère les articles d'une catégorie spécifique
   * GET /api/articles/category/{categorySlug}?page=0&size=5&sort=publishedAt,desc
   */
  async getArticlesByCategory(
    categorySlug: string,
    page: number = 0,
    size: number = 5,
    sort: string = 'publishedAt,desc'
  ): Promise<Page<ArticleResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        size: size.toString(),
        sort,
      });

      const articlesPage = await get<Page<ArticleResponse>>(
        `/api/articles/category/${categorySlug}?${params}`
      );
      
      console.log('✅ Articles by category retrieved:', {
        category: categorySlug,
        totalElements: articlesPage.totalElements,
        currentPage: articlesPage.number,
      });
      
      return articlesPage;
      
    } catch (error) {
      console.error(`❌ Failed to get articles for category ${categorySlug}:`, error);
      throw error;
    }
  },

  // ---------------------------------------------------------------------------
  // ENDPOINTS SÉCURISÉS (ÉDITEURS + ADMINS)
  // ---------------------------------------------------------------------------

  /**
   * Crée un nouvel article en brouillon
   * POST /api/articles
   * Requiert : EDITEUR ou ADMINISTRATEUR
   */
  async createArticle(articleData: ArticleRequest): Promise<ArticleResponse> {
    try {
      const article = await post<ArticleResponse>('/api/articles', articleData);
      console.log('✅ Article created:', article.title, '(Status: DRAFT)');
      return article;
      
    } catch (error) {
      console.error('❌ Failed to create article:', error);
      throw error;
    }
  },

  /**
   * Met à jour un article existant
   * PUT /api/articles/{id}
   * Requiert : EDITEUR (propre article) ou ADMINISTRATEUR (tous)
   */
  async updateArticle(id: string, articleData: ArticleRequest): Promise<ArticleResponse> {
    try {
      const article = await put<ArticleResponse>(`/api/articles/${id}`, articleData);
      console.log('✅ Article updated:', article.title);
      return article;
      
    } catch (error) {
      console.error(`❌ Failed to update article ${id}:`, error);
      throw error;
    }
  },

  /**
   * Publie un article (change le statut de DRAFT à PUBLISHED)
   * POST /api/articles/{id}/publish
   * Requiert : EDITEUR ou ADMINISTRATEUR
   */
  async publishArticle(id: string): Promise<ArticleResponse> {
    try {
      const article = await post<ArticleResponse>(`/api/articles/${id}/publish`);
      console.log('✅ Article published:', article.title);
      return article;
      
    } catch (error) {
      console.error(`❌ Failed to publish article ${id}:`, error);
      throw error;
    }
  },

  /**
   * Archive un article (change le statut à ARCHIVED)
   * POST /api/articles/{id}/archive
   * Requiert : EDITEUR ou ADMINISTRATEUR
   */
  async archiveArticle(id: string): Promise<ArticleResponse> {
    try {
      const article = await post<ArticleResponse>(`/api/articles/${id}/archive`);
      console.log('✅ Article archived:', article.title);
      return article;
      
    } catch (error) {
      console.error(`❌ Failed to archive article ${id}:`, error);
      throw error;
    }
  },

  /**
   * Supprime définitivement un article
   * DELETE /api/articles/{id}
   * Requiert : ADMINISTRATEUR uniquement
   */
  async deleteArticle(id: string): Promise<void> {
    try {
      await del(`/api/articles/${id}`);
      console.log('✅ Article deleted permanently:', id);
      
    } catch (error) {
      console.error(`❌ Failed to delete article ${id}:`, error);
      throw error;
    }
  },

  // ---------------------------------------------------------------------------
  // MÉTHODES UTILITAIRES ET RECHERCHE
  // ---------------------------------------------------------------------------

  /**
   * Recherche d'articles avec filtres multiples
   * Utilise l'endpoint de recherche du backend si disponible
   */
  async searchArticles(searchParams: SearchParams): Promise<Page<ArticleResponse>> {
    try {
      const params = new URLSearchParams();
      
      if (searchParams.query) params.append('q', searchParams.query);
      if (searchParams.categoryId) params.append('categoryId', searchParams.categoryId);
      if (searchParams.status) params.append('status', searchParams.status);
      if (searchParams.page !== undefined) params.append('page', searchParams.page.toString());
      if (searchParams.size !== undefined) params.append('size', searchParams.size.toString());
      if (searchParams.sort) params.append('sort', searchParams.sort);

      const articlesPage = await get<Page<ArticleResponse>>(
        `/api/articles/search?${params}`
      );
      
      console.log('✅ Articles search completed:', {
        query: searchParams.query,
        totalResults: articlesPage.totalElements,
      });
      
      return articlesPage;
      
    } catch (error) {
      console.error('❌ Failed to search articles:', error);
      throw error;
    }
  },

  /**
   * Récupère les articles d'un auteur spécifique (pour éditeurs)
   * Utilise les endpoints existants avec filtres
   */
  async getArticlesByAuthor(
    authorId: string,
    page: number = 0,
    size: number = 10
  ): Promise<Page<ArticleResponse>> {
    try {
      const params = new URLSearchParams({
        authorId,
        page: page.toString(),
        size: size.toString(),
        sort: 'updatedAt,desc',
      });

      const articlesPage = await get<Page<ArticleResponse>>(
        `/api/articles?${params}`
      );
      
      console.log('✅ Articles by author retrieved:', {
        authorId,
        totalElements: articlesPage.totalElements,
      });
      
      return articlesPage;
      
    } catch (error) {
      console.error(`❌ Failed to get articles by author ${authorId}:`, error);
      throw error;
    }
  },

  /**
   * Récupère les articles par statut (pour éditeurs/admins)
   */
  async getArticlesByStatus(
    status: ArticleStatus,
    page: number = 0,
    size: number = 10
  ): Promise<Page<ArticleResponse>> {
    try {
      const params = new URLSearchParams({
        status,
        page: page.toString(),
        size: size.toString(),
        sort: 'updatedAt,desc',
      });

      const articlesPage = await get<Page<ArticleResponse>>(
        `/api/articles?${params}`
      );
      
      console.log('✅ Articles by status retrieved:', {
        status,
        totalElements: articlesPage.totalElements,
      });
      
      return articlesPage;
      
    } catch (error) {
      console.error(`❌ Failed to get articles by status ${status}:`, error);
      throw error;
    }
  },

  /**
   * Récupère les brouillons de l'utilisateur connecté
   */
  async getMyDrafts(page: number = 0, size: number = 10): Promise<Page<ArticleResponse>> {
    return this.getArticlesByStatus('DRAFT', page, size);
  },

  /**
   * Récupère tous les articles publiés de l'utilisateur connecté
   */
  async getMyPublishedArticles(page: number = 0, size: number = 10): Promise<Page<ArticleResponse>> {
    return this.getArticlesByStatus('PUBLISHED', page, size);
  },

  /**
   * Récupère les statistiques des articles (pour dashboard)
   */
  async getArticleStats(): Promise<{
    totalArticles: number;
    publishedArticles: number;
    draftArticles: number;
    archivedArticles: number;
  }> {
    try {
      const stats = await get<{
        totalArticles: number;
        publishedArticles: number;
        draftArticles: number;
        archivedArticles: number;
      }>('/api/articles/stats');
      
      console.log('✅ Article stats retrieved:', stats);
      return stats;
      
    } catch (error) {
      console.error('❌ Failed to get article stats:', error);
      throw error;
    }
  },
}; 

File: backend/src/main/java/com/newsplatform/service/ArticleService.java
package com.newsplatform.service;

import com.newsplatform.entity.Article;
import com.newsplatform.entity.ArticleStatus;
import com.newsplatform.entity.Category;
import com.newsplatform.entity.User;
import com.newsplatform.exception.BusinessException;
import com.newsplatform.exception.ResourceNotFoundException;
import com.newsplatform.exception.ValidationException;
import com.newsplatform.repository.ArticleRepository;
import com.newsplatform.repository.CategoryRepository;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Service pour la gestion des articles selon les principes DDD.
 * Couche Service : Orchestration de la logique métier et validation
 * 
 * Responsabilités :
 * - Orchestration des opérations CRUD avec validation métier
 * - Gestion des autorisations selon les rôles utilisateur
 * - Coordination entre les repositories
 * - Application des règles métier complexes
 * - Transformation et validation des données
 * 
 * @author Équipe Développement
 * @version 2.0 - Refactoring DDD complet
 */
@Service
@Transactional
public class ArticleService {
    
    private static final Logger logger = LoggerFactory.getLogger(ArticleService.class);
    
    private final ArticleRepository articleRepository;
    private final CategoryRepository categoryRepository;
    
    @Autowired
    public ArticleService(ArticleRepository articleRepository, CategoryRepository categoryRepository) {
        this.articleRepository = articleRepository;
        this.categoryRepository = categoryRepository;
    }

    // ===============================================
    // REQUÊTES PUBLIQUES (ARTICLES PUBLIÉS)
    // ===============================================
    
    /**
     * Récupère les derniers articles publiés pour la page d'accueil.
     * Optimisé avec limite fixe pour les performances.
     * 
     * @param limit nombre maximum d'articles (par défaut 10)
     * @return liste des derniers articles publiés
     * @throws BusinessException en cas d'erreur technique
     */
    @Transactional(readOnly = true)
    public List<Article> getRecentPublishedArticles(int limit) {
        // Validation des paramètres
        if (limit <= 0 || limit > 50) {
            throw new ValidationException("La limite doit être entre 1 et 50 articles");
        }
        
        try {
            Pageable pageable = PageRequest.of(0, limit);
            return articleRepository.findTopRecentPublished(ArticleStatus.PUBLISHED, pageable);
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la récupération des articles récents", e);
        }
    }

    /**
     * Récupère les derniers articles publiés avec la limite par défaut.
     * Méthode de convenance pour compatibilité.
     * 
     * @return liste des 10 derniers articles publiés
     */
    @Transactional(readOnly = true)
    public List<Article> getRecentArticles() {
        return getRecentPublishedArticles(10);
    }
    
    /**
     * Récupère un article publié par son slug pour affichage public.
     * Sécurité : seuls les articles publiés sont accessibles.
     * 
     * @param slug slug unique de l'article
     * @return article publié
     * @throws ResourceNotFoundException si l'article n'existe pas ou n'est pas publié
     */
    @Transactional(readOnly = true)
    public Article getPublishedArticleBySlug(String slug) {
        // Validation des paramètres
        if (slug == null || slug.trim().isEmpty()) {
            throw new ValidationException("Le slug ne peut pas être vide");
        }
        
        return articleRepository.findPublishedBySlug(slug.trim(), ArticleStatus.PUBLISHED)
                .orElseThrow(() -> new ResourceNotFoundException(
                    "Article publié non trouvé avec le slug : " + slug));
    }
    
    /**
     * Récupère les articles publiés avec pagination.
     * Optimisé avec JOIN FETCH pour éviter N+1.
     * 
     * @param pageable configuration de pagination
     * @return page d'articles publiés
     * @throws ValidationException si pageable est invalide
     */
    @Transactional(readOnly = true)
    public Page<Article> getPublishedArticles(Pageable pageable) {
        // Validation des paramètres
        validatePageable(pageable);
        
        try {
            return articleRepository.findRecentPublishedArticles(ArticleStatus.PUBLISHED, pageable);
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la récupération des articles paginés", e);
        }
    }
    
    /**
     * Récupère les articles publiés d'une catégorie avec pagination.
     * 
     * @param categorySlug slug de la catégorie
     * @param pageable configuration de pagination
     * @return page d'articles de la catégorie
     * @throws ResourceNotFoundException si la catégorie n'existe pas
     */
    @Transactional(readOnly = true)
    public Page<Article> getPublishedArticlesByCategory(String categorySlug, Pageable pageable) {
        // Validation des paramètres
        if (categorySlug == null || categorySlug.trim().isEmpty()) {
            throw new ValidationException("Le slug de catégorie ne peut pas être vide");
        }
        validatePageable(pageable);
        
        // Récupération de la catégorie
        Category category = categoryRepository.findBySlug(categorySlug.trim())
                .orElseThrow(() -> new ResourceNotFoundException(
                    "Catégorie non trouvée avec le slug : " + categorySlug));
        
        try {
            return articleRepository.findPublishedByCategory(category, ArticleStatus.PUBLISHED, pageable);
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la récupération des articles par catégorie", e);
        }
    }
    
    /**
     * Recherche d'articles publiés par terme de recherche.
     * 
     * @param searchTerm terme de recherche
     * @param pageable configuration de pagination
     * @return page d'articles correspondants
     */
    @Transactional(readOnly = true)
    public Page<Article> searchPublishedArticles(String searchTerm, Pageable pageable) {
        // Validation des paramètres
        if (searchTerm == null) {
            searchTerm = "";
        }
        validatePageable(pageable);
        
        try {
            return articleRepository.searchPublishedArticles(
                searchTerm.trim(), 
                ArticleStatus.PUBLISHED, 
                pageable
            );
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la recherche d'articles", e);
        }
    }

    // ===============================================
    // REQUÊTES D'ADMINISTRATION
    // ===============================================
    
    /**
     * Récupère un article par son ID (tous statuts) pour administration.
     * Utilisé par les éditeurs et administrateurs.
     * 
     * @param id identifiant unique de l'article
     * @return article trouvé
     * @throws ResourceNotFoundException si l'article n'existe pas
     */
    @Transactional(readOnly = true)
    public Article getArticleById(UUID id) {
        // Validation des paramètres
        if (id == null) {
            throw new ValidationException("L'ID de l'article ne peut pas être null");
        }
        
        return articleRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException(
                    "Article non trouvé avec l'ID : " + id));
    }
    
    /**
     * Récupère un article par son slug (tous statuts) pour administration.
     * 
     * @param slug slug unique de l'article
     * @return article trouvé
     * @throws ResourceNotFoundException si l'article n'existe pas
     */
    @Transactional(readOnly = true)
    public Article getArticleBySlugForAdmin(String slug) {
        // Validation des paramètres
        if (slug == null || slug.trim().isEmpty()) {
            throw new ValidationException("Le slug ne peut pas être vide");
        }
        
        return articleRepository.findBySlugWithDetails(slug.trim())
                .orElseThrow(() -> new ResourceNotFoundException(
                    "Article non trouvé avec le slug : " + slug));
    }
    
    /**
     * Récupère tous les articles avec pagination pour administration.
     * 
     * @param pageable configuration de pagination
     * @return page de tous les articles
     */
    @Transactional(readOnly = true)
    public Page<Article> getAllArticles(Pageable pageable) {
        validatePageable(pageable);
        
        try {
            return articleRepository.findAllWithDetails(pageable);
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la récupération de tous les articles", e);
        }
    }
    
    /**
     * Récupère les articles d'un auteur avec pagination.
     * 
     * @param author auteur des articles
     * @param pageable configuration de pagination
     * @return page des articles de l'auteur
     */
    @Transactional(readOnly = true)
    public Page<Article> getArticlesByAuthor(User author, Pageable pageable) {
        // Validation des paramètres
        if (author == null) {
            throw new ValidationException("L'auteur ne peut pas être null");
        }
        validatePageable(pageable);
        
        try {
            return articleRepository.findByAuthor(author, pageable);
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la récupération des articles par auteur", e);
        }
    }
    
    /**
     * Récupère les articles par statut avec pagination.
     * 
     * @param status statut des articles
     * @param pageable configuration de pagination
     * @return page des articles avec le statut donné
     */
    @Transactional(readOnly = true)
    public Page<Article> getArticlesByStatus(ArticleStatus status, Pageable pageable) {
        // Validation des paramètres
        if (status == null) {
            throw new ValidationException("Le statut ne peut pas être null");
        }
        validatePageable(pageable);
        
        try {
            return articleRepository.findByStatus(status, pageable);
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la récupération des articles par statut", e);
        }
    }

    /**
     * Récupère les articles d'un auteur avec un statut spécifique et pagination.
     * 
     * @param author auteur des articles
     * @param status statut des articles
     * @param pageable configuration de pagination
     * @return page des articles de l'auteur avec le statut donné
     */
    @Transactional(readOnly = true)
    public Page<Article> getArticlesByAuthorAndStatus(User author, ArticleStatus status, Pageable pageable) {
        // Validation des paramètres
        if (author == null) {
            throw new ValidationException("L'auteur ne peut pas être null");
        }
        if (status == null) {
            throw new ValidationException("Le statut ne peut pas être null");
        }
        validatePageable(pageable);
        
        try {
            return articleRepository.findByAuthorAndStatus(author, status, pageable);
        } catch (Exception e) {
            throw new BusinessException("Erreur lors de la récupération des articles par auteur et statut", e);
        }
    }

    // ===============================================
    // OPÉRATIONS CRUD ET GESTION DES STATUTS
    // ===============================================
    
    /**
     * Crée un nouvel article en brouillon.
     * Validation complète selon les règles métier DDD.
     * 
     * @param title titre de l'article
     * @param content contenu de l'article
     * @param categorySlug slug de la catégorie
     * @param author auteur de l'article
     * @return article créé
     * @throws ValidationException si les données sont invalides
     * @throws ResourceNotFoundException si la catégorie n'existe pas
     */
    @Transactional
    public Article createArticle(String title, String content, String categorySlug, User author) {
        logger.info("📝 Création d'un nouvel article - Titre: '{}', Catégorie: '{}', Auteur: '{}'", 
                   title != null ? title.substring(0, Math.min(title.length(), 50)) + "..." : "null",
                   categorySlug, 
                   author != null ? author.getUsername() : "null");
        
        // Validation des paramètres
        validateArticleCreationParams(title, content, categorySlug, author);
        
        // Récupération et validation de la catégorie
        Category category = categoryRepository.findBySlug(categorySlug.trim())
                .orElseThrow(() -> {
                    logger.warn("⚠️ Tentative de création d'article avec catégorie inexistante: '{}'", categorySlug);
                    return new ResourceNotFoundException("Catégorie non trouvée avec le slug : " + categorySlug);
                });
        
        try {
            // Création avec constructeur DDD (automatiquement en DRAFT)
            Article article = new Article(title.trim(), content.trim(), category, author);
            
            // Validation d'unicité du slug
            if (articleRepository.existsBySlug(article.getSlug())) {
                logger.warn("⚠️ Tentative de création d'article avec slug déjà existant: '{}'", article.getSlug());
                throw new ValidationException("Un article avec ce slug existe déjà : " + article.getSlug());
            }
            
            Article savedArticle = articleRepository.save(article);
            logger.info("✅ Article créé avec succès - ID: {}, Slug: '{}'", savedArticle.getId(), savedArticle.getSlug());
            
            return savedArticle;
        } catch (Exception e) {
            if (e instanceof ValidationException || e instanceof ResourceNotFoundException) {
                throw e;
            }
            logger.error("❌ Erreur lors de la création de l'article - Titre: '{}', Auteur: '{}'", 
                        title, author != null ? author.getUsername() : "null", e);
            throw new BusinessException("Erreur lors de la création de l'article", e);
        }
    }
    
    /**
     * Met à jour le contenu d'un article existant.
     * Utilise la méthode métier DDD updateContent().
     * 
     * @param articleId ID de l'article à modifier
     * @param newTitle nouveau titre
     * @param newContent nouveau contenu
     * @param currentUser utilisateur effectuant la modification
     * @return article mis à jour
     * @throws ResourceNotFoundException si l'article n'existe pas
     * @throws ValidationException si les données sont invalides
     * @throws BusinessException si l'utilisateur n'a pas les droits
     */
    @Transactional
    public Article updateArticleContent(UUID articleId, String newTitle, String newContent, User currentUser) {
        // Validation des paramètres
        if (articleId == null) {
            throw new ValidationException("L'ID de l'article ne peut pas être null");
        }
        validateArticleContentParams(newTitle, newContent);
        if (currentUser == null) {
            throw new ValidationException("L'utilisateur actuel ne peut pas être null");
        }
        
        // Récupération de l'article
        Article article = getArticleById(articleId);
        
        // Vérification des droits (seul l'auteur ou un admin peut modifier)
        if (!article.getAuthor().equals(currentUser) && !currentUser.getRole().equals(User.UserRole.ADMINISTRATEUR)) {
            throw new BusinessException("Vous n'êtes pas autorisé à modifier cet article");
        }
        
        try {
            // Validation d'unicité du slug si le titre change
            String potentialSlug = generateSlugFromTitle(newTitle.trim());
            if (!potentialSlug.equals(article.getSlug()) && 
                articleRepository.existsBySlugAndIdNot(potentialSlug, articleId)) {
                throw new ValidationException("Un article avec ce titre (slug) existe déjà");
            }
            
            // Utilisation de la méthode métier DDD
            article.updateContent(newTitle.trim(), newContent.trim(), null);
            
            return articleRepository.save(article);
        } catch (Exception e) {
            if (e instanceof ValidationException || e instanceof BusinessException) {
                throw e;
            }
            throw new BusinessException("Erreur lors de la mise à jour de l'article", e);
        }
    }
    
    /**
     * Publie un article en brouillon.
     * Utilise la méthode métier DDD publish().
     * 
     * @param articleId ID de l'article à publier
     * @param currentUser utilisateur effectuant l'action
     * @return article publié
     * @throws ResourceNotFoundException si l'article n'existe pas
     * @throws BusinessException si l'article ne peut pas être publié
     */
    @Transactional
    public Article publishArticle(UUID articleId, User currentUser) {
        logger.info("📢 Tentative de publication d'article - ID: {}, Utilisateur: '{}'", 
                   articleId, currentUser != null ? currentUser.getUsername() : "null");
        
        // Validation des paramètres
        if (articleId == null) {
            throw new ValidationException("L'ID de l'article ne peut pas être null");
        }
        if (currentUser == null) {
            throw new ValidationException("L'utilisateur actuel ne peut pas être null");
        }
        
        // Récupération de l'article
        Article article = getArticleById(articleId);
        
        // Vérification des droits
        if (!article.getAuthor().equals(currentUser) && !currentUser.getRole().equals(User.UserRole.ADMINISTRATEUR)) {
            logger.warn("🚫 Tentative de publication non autorisée - Article: {}, Auteur: '{}', Utilisateur tentant: '{}'", 
                       articleId, article.getAuthor().getUsername(), currentUser.getUsername());
            throw new BusinessException("Vous n'êtes pas autorisé à publier cet article");
        }
        
        try {
            logger.debug("🔄 Publication en cours - Article: '{}' (Status actuel: {})", 
                        article.getTitle(), article.getStatus());
            
            // Utilisation de la méthode métier DDD
            article.publish();
            
            Article publishedArticle = articleRepository.save(article);
            logger.info("✅ Article publié avec succès - ID: {}, Titre: '{}', Publié par: '{}'", 
                       publishedArticle.getId(), 
                       publishedArticle.getTitle().substring(0, Math.min(publishedArticle.getTitle().length(), 50)) + "...", 
                       currentUser.getUsername());
            
            return publishedArticle;
        } catch (Exception e) {
            if (e instanceof ValidationException || e instanceof BusinessException) {
                throw e;
            }
            logger.error("❌ Erreur lors de la publication de l'article - ID: {}, Utilisateur: '{}'", 
                        articleId, currentUser.getUsername(), e);
            throw new BusinessException("Erreur lors de la publication de l'article", e);
        }
    }
    
    /**
     * Archive un article publié.
     * Utilise la méthode métier DDD archive().
     * 
     * @param articleId ID de l'article à archiver
     * @param currentUser utilisateur effectuant l'action
     * @return article archivé
     */
    @Transactional
    public Article archiveArticle(UUID articleId, User currentUser) {
        // Validation des paramètres
        if (articleId == null) {
            throw new ValidationException("L'ID de l'article ne peut pas être null");
        }
        if (currentUser == null) {
            throw new ValidationException("L'utilisateur actuel ne peut pas être null");
        }
        
        // Récupération de l'article
        Article article = getArticleById(articleId);
        
        // Vérification des droits
        if (!article.getAuthor().equals(currentUser) && !currentUser.getRole().equals(User.UserRole.ADMINISTRATEUR)) {
            throw new BusinessException("Vous n'êtes pas autorisé à archiver cet article");
        }
        
        try {
            // Utilisation de la méthode métier DDD
            article.archive();
            
            return articleRepository.save(article);
        } catch (Exception e) {
            if (e instanceof ValidationException || e instanceof BusinessException) {
                throw e;
            }
            throw new BusinessException("Erreur lors de l'archivage de l'article", e);
        }
    }

    // ===============================================
    // STATISTIQUES ET MÉTRIQUES
    // ===============================================
    
    /**
     * Compte le nombre d'articles publiés d'une catégorie.
     * 
     * @param categorySlug slug de la catégorie
     * @return nombre d'articles publiés
     */
    @Transactional(readOnly = true)
    public long countPublishedArticlesByCategory(String categorySlug) {
        if (categorySlug == null || categorySlug.trim().isEmpty()) {
            throw new ValidationException("Le slug de catégorie ne peut pas être vide");
        }
        
        Category category = categoryRepository.findBySlug(categorySlug.trim())
                .orElseThrow(() -> new ResourceNotFoundException(
                    "Catégorie non trouvée avec le slug : " + categorySlug));
        
        return articleRepository.countPublishedByCategory(category, ArticleStatus.PUBLISHED);
    }
    
    /**
     * Compte le nombre total d'articles publiés.
     * 
     * @return nombre total d'articles publiés
     */
    @Transactional(readOnly = true)
    public long countTotalPublishedArticles() {
        return articleRepository.countPublished(ArticleStatus.PUBLISHED);
    }
    
    /**
     * Compte les articles d'un auteur.
     * 
     * @param author auteur des articles
     * @return nombre d'articles de l'auteur
     */
    @Transactional(readOnly = true)
    public long countArticlesByAuthor(User author) {
        if (author == null) {
            throw new ValidationException("L'auteur ne peut pas être null");
        }
        
        return articleRepository.countByAuthor(author);
    }

    // ===============================================
    // MÉTHODES UTILITAIRES PRIVÉES
    // ===============================================
    
    /**
     * Valide les paramètres de pagination.
     * 
     * @param pageable objet de pagination à valider
     * @throws ValidationException si invalide
     */
    private void validatePageable(Pageable pageable) {
        if (pageable == null) {
            throw new ValidationException("Les paramètres de pagination ne peuvent pas être null");
        }
        if (pageable.getPageSize() > 100) {
            throw new ValidationException("La taille de page ne peut pas dépasser 100 éléments");
        }
    }
    
    /**
     * Valide les paramètres de création d'article.
     */
    private void validateArticleCreationParams(String title, String content, String categorySlug, User author) {
        if (title == null || title.trim().isEmpty()) {
            throw new ValidationException("Le titre ne peut pas être vide");
        }
        if (content == null || content.trim().isEmpty()) {
            throw new ValidationException("Le contenu ne peut pas être vide");
        }
        if (categorySlug == null || categorySlug.trim().isEmpty()) {
            throw new ValidationException("Le slug de catégorie ne peut pas être vide");
        }
        if (author == null) {
            throw new ValidationException("L'auteur ne peut pas être null");
        }
    }
    
    /**
     * Valide les paramètres de contenu d'article.
     */
    private void validateArticleContentParams(String title, String content) {
        if (title == null || title.trim().isEmpty()) {
            throw new ValidationException("Le titre ne peut pas être vide");
        }
        if (content == null || content.trim().isEmpty()) {
            throw new ValidationException("Le contenu ne peut pas être vide");
        }
    }
    
    /**
     * Génère un slug à partir d'un titre.
     * Méthode utilitaire pour validation d'unicité.
     */
    private String generateSlugFromTitle(String title) {
        return title.toLowerCase()
                   .replaceAll("[^a-z0-9\\s-]", "")
                   .replaceAll("\\s+", "-")
                   .replaceAll("-+", "-")
                   .replaceAll("^-|-$", "");
    }
}
